### ISSUE TITLE
[bug?] old event published on each card swipe

### ISSUE THREAD

**Issue by wuub** [uhppoted/uhppoted-mqtt#15]
I'm not sure if this is a configuration issue on my end, but every time I swipe a card or enter a code uhppoted-mqtt first sends the previous event that it saw and only after that the current event (look at timestamps and receive dates).

the event being resent can be hours old. 

No other mqtt topic shows this behavior.

Version: uhppoted-mqtt_v0.8.6/linux

```
2023/10/28 12:47:16 [client]   enter Publish
2023/10/28 12:47:16 [client]   sending publish message, topic: wiarus/gateway/events
2023/10/28 12:47:16 [net]      obound msg to write 26
 ... sent 64 bytes to 192.168.1.123:60000

2023/10/28 12:47:16 [net]      obound wrote msg, id: 26
2023/10/28 12:47:16 [net]      outgoing waiting for an outbound message
 ... request
 ...          00000000  17 b0 00 00 6a 08 4e 0d  d1 00 00 00 00 00 00 00  |....j.N.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          

 ... received 64 bytes from 192.168.1.123:60000
 ... response
 ...          00000000  17 b0 00 00 6a 08 4e 0d  d1 00 00 00 01 01 01 01  |....j.N.........|
 ...          00000010  6f 5a 25 00 20 23 10 28  12 47 11 01 00 00 00 00  |oZ%. #.(.G......|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          
2023/10/28 12:47:16 [client]   enter Publish
2023/10/28 12:47:16 [client]   sending publish message, topic: wiarus/gateway/events
2023/10/28 12:47:16 [net]      obound msg to write 27
2023/10/28 12:47:16 [net]      obound wrote msg, id: 27
2023/10/28 12:47:16 [net]      outgoing waiting for an outbound message
2023/10/28 12:47:16 [net]      startIncoming Received Message
2023/10/28 12:47:16 [net]      startIncomingComms: got msg on ibound
2023/10/28 12:47:16 [store]    memorystore del: message 26 was deleted
2023/10/28 12:47:16 [net]      startIncomingComms: received puback, id: 26
2023/10/28 12:47:16 [net]      logic waiting for msg on ibound
2023/10/28 12:47:16 [net]      startIncoming Received Message
2023/10/28 12:47:16 [net]      startIncomingComms: got msg on ibound
2023/10/28 12:47:16 [store]    memorystore del: message 27 was deleted
2023/10/28 12:47:16 [net]      startIncomingComms: received puback, id: 27
2023/10/28 12:47:16 [net]      logic waiting for msg on ibound
2023/10/28 12:47:18 INFO   watchdog     OK
2023/10/28 12:47:19 [pinger]   ping check 2.858216454
2023/10/28 12:47:23 INFO   watchdog     OK
2023/10/28 12:47:24 [pinger]   ping check 7.857553456
2023/10/28 12:47:28 INFO   watchdog     OK

```

![image](https://github.com/uhppoted/uhppoted-mqtt/assets/588857/5fb3ba39-abac-40eb-bb85-41c66af4fbe4)

**Comment by uhppoted** [comment]
Hmmm, weird - it doesn't do this on my system and I'm reasonably sure it doesn't send the previous event so it's most likely something to do with the broker configuration. I'm running _HiveMQ_  - presumably you're running _mosquito_ ?

It **may** be something to do with retained messages (defaults to `true`) which doesn't always do what you would expect. Does the problem still appear if you disable _retained messages_ in the _uhppoted.conf_ file?
```
mqtt.alerts.retained = false
```
There's also a note in the [FAQ](https://github.com/uhppoted/uhppoted-mqtt/blob/main/documentation/FAQ.md) about clearing old messages:
```
2. Why am I still getting old events even though I have set mqtt.alerts.retained = false ?

The MQTT broker probably still has retained events from when mqtt.alerts.retained was set to true.
You need to manually clear the events - normally by sending an empty retained message to the topics:

mqtt publish --topic 'uhppoted/gateway/events' --message '' -r
mqtt publish --topic 'uhppoted/gateway/system' --message '' -r
```

**Comment by wuub** [comment]
1. I turned off retention, completely wiped mosquitto installation and the doubled events persisted.
2. I kept retention off and used hivemq wss:// broker

**Comment by wuub** [comment]
This is what I see with debug on for repeated messages 

```
... listening
2023/10/30 14:35:45 INFO   event        223217770  true  true  false false false false false false 0    2023-10-30 15:35:40 0          0 01 00 | 247     1   true  1 1     30XXXX     2023-10-30 15:35:40 1
 ... sent 64 bytes to 192.168.1.123:60000

 ... request
 ...          00000000  17 b0 00 00 6a 08 4e 0d  00 00 00 00 00 00 00 00  |....j.N.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          

 ... received 64 bytes from 192.168.1.123:60000
 ... response
 ...          00000000  17 b0 00 00 6a 08 4e 0d  01 00 00 00 02 01 01 01  |....j.N.........|
 ...          00000010  01 00 00 00 20 23 05 13  08 15 18 2c 00 00 00 00  |.... #.....,....|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          
 ... sent 64 bytes to 192.168.1.123:60000

 ... request
 ...          00000000  17 b0 00 00 6a 08 4e 0d  ff ff ff ff 00 00 00 00  |....j.N.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          

 ... received 64 bytes from 192.168.1.123:60000
 ... response
 ...          00000000  17 b0 00 00 6a 08 4e 0d  f7 00 00 00 01 01 01 01  |....j.N.........|
 ...          00000010  81 b2 04 00 20 23 10 30  15 35 40 01 00 00 00 00  |.... #.0.5@.....|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          
 ... sent 64 bytes to 192.168.1.123:60000

 ... request
 ...          00000000  17 b0 00 00 6a 08 4e 0d  f6 00 00 00 00 00 00 00  |....j.N.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          

 ... received 64 bytes from 192.168.1.123:60000
 ... response
 ...          00000000  17 b0 00 00 6a 08 4e 0d  f6 00 00 00 01 01 01 01  |....j.N.........|
 ...          00000010  81 b2 04 00 20 23 10 30  15 34 25 01 00 00 00 00  |.... #.0.4%.....|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          
2023/10/30 14:35:45 INFO   mqttd        got event {%!s(uint32=223217770) %!s(uint32=246) %!s(uint8=1) %!s(bool=true) %!s(uint8=1) %!s(uint8=1) %!s(uint32=30XXXX) 2023-10-30 15:34:25 %!s(uint8=1)}
2023/10/30 14:35:45 INFO   mqttd        sending message to wiarus/gateway/events -> {"message":{"event":{"event":{"access-granted":true,"card-number":30XXXX,"device-id":223217770,"direction":1,"direction-text":"in","door-id":1,"event-id":246,"event-reason":1,"event-reason-text":"swipe","event-type":1,"event-type-text":"card swipe","timestamp":"2023-10-30 15:34:25 CET"}}},"hmac":"9760a08fb5f23e2243a0a3f4f2ecf1e8f1e5792600a3d25d6894a89f97fa537f"}
2023/10/30 14:35:45 [client]   enter Publish
2023/10/30 14:35:45 [client]   sending publish message, topic: wiarus/gateway/events
2023/10/30 14:35:45 [net]      obound msg to write 2
 ... sent 64 bytes to 192.168.1.123:60000

 ... request
 ...          00000000  17 b0 00 00 6a 08 4e 0d  f7 00 00 00 00 00 00 00  |....j.N.........|
 ...          00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          

2023/10/30 14:35:45 [net]      obound wrote msg, id: 2
2023/10/30 14:35:45 [net]      outgoing waiting for an outbound message
 ... received 64 bytes from 192.168.1.123:60000
 ... response
 ...          00000000  17 b0 00 00 6a 08 4e 0d  f7 00 00 00 01 01 01 01  |....j.N.........|
 ...          00000010  81 b2 04 00 20 23 10 30  15 35 40 01 00 00 00 00  |.... #.0.5@.....|
 ...          00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 ...          
2023/10/30 14:35:45 INFO   mqttd        got event {%!s(uint32=223217770) %!s(uint32=247) %!s(uint8=1) %!s(bool=true) %!s(uint8=1) %!s(uint8=1) %!s(uint32=30XXXX) 2023-10-30 15:35:40 %!s(uint8=1)}
2023/10/30 14:35:45 INFO   mqttd        sending message to wiarus/gateway/events -> {"message":{"event":{"event":{"access-granted":true,"card-number":30XXXX,"device-id":223217770,"direction":1,"direction-text":"in","door-id":1,"event-id":247,"event-reason":1,"event-reason-text":"swipe","event-type":1,"event-type-text":"card swipe","timestamp":"2023-10-30 15:35:40 CET"}}},"hmac":"1faa3dea3bb8e7c9ea77c314069fdcf195f640951457c0bf9c14d7b58f73f0a0"}
2023/10/30 14:35:45 [client]   enter Publish
2023/10/30 14:35:45 [client]   sending publish message, topic: wiarus/gateway/events
2023/10/30 14:35:45 [net]      obound msg to write 3
2023/10/30 14:35:45 [net]      obound wrote msg, id: 3
2023/10/30 14:35:45 [net]      outgoing waiting for an outbound message
```

**Comment by wuub** [comment]
If i had to guess I would say that this map is somehow not being updated https://github.com/uhppoted/uhppoted-lib/blob/master/uhppoted/listen.go#L159C3-L159C41 

causing the lib to refetch the current-1 event every time.

but it's just a guess :)

**Comment by uhppoted** [comment]
Wow! Nice find .. thanks for looking into it! 

First thought is that the _EventMap_ doesn't have any locking around it (very old code :-( ) so it's not safe for concurrent access. But let me do some digging and see if I can figure out what on earth is happening...

**Comment by uhppoted** [comment]
Well, that's embarrassing :-(.

It is in fact manifesting on my system but for complicated reasons (basically an unexpected interaction between _Docker_ and the _mqtt.events.retrieved_ file) I simply wasn't seeing it.

The cause this line [listen.go:155](https://github.com/uhppoted/uhppoted-lib/blob/master/uhppoted/listen.go#L155):
```
        ...
	retrieved, ok := received.retrieved[deviceID]
	if ok && retrieved != uint32(last) {
		first = retrieved
	}
        ...
```
should be:
```
        ...
	retrieved, ok := received.retrieved[deviceID]
	if ok && retrieved != uint32(last) {
		first = retrieved + 1
	}
        ...
```
So, yeah, you could probably call that an _off-by-one_ error :-).

I just need to dig a bit further because the event listener is used in a couple of other places and there might be an obscure reason why it's like that - but if nothing turns up I'll push a fix to _github_ in the next day or so - if it's causing a problem though let me know.

And thanks for the detailed debugging info - made finding it a lot easier!

**Comment by wuub** [comment]
Thank you so much for looking into it :) There's nothing embarrassing about it at all. I'll apply that patch locally and see what/if anything blows up. 

On my end there's no rush as far as official release goes. I can run from a local build no problem :)

**Comment by uhppoted** [comment]
Hi,

Just a quick note to let you know I've pushed the fix to the _github_ repository - it turned out to be easy because all the other dependencies had all been migrated off over the last year or so. 

You'll have to clone the repo and build locally to use it - the fix will be in the next release but that's probably only going to happen at the end of November. Let me know if you run into anything weird with it..

And .. just a quick heads up: I'm going to rework the events handling in _uhppoted-mqtt_ because it's **less** than intuitive. Basically if _uhppoted-mqtt_ has been offline for a while, it won't publish swipe events to the _events_ topic until it has caught up the backlog. Which is not .. ideal. However, that's a separate task and (hopefully) shouldn't affect you - the details are here if you're curious: [Issue#16:Rework events handling](https://github.com/uhppoted/uhppoted-mqtt/issues/16).

**Comment by uhppoted** [comment]
Hi,

Am going to close this out (the fix will go out in the next release) - feel free to reopen it if you need to :-). And thanks again for the high quality bug report - made things **so** much easier..

**Comment by uhppoted** [comment]
Just a quick note - I've reworked the _uhppoted-mqtt_ events handling to add a _live_ events topic that publishes controller events as they are received. The existing _events_ topic also publishes events that occurred while _uhppoted-mqtt_ service was offline and if the events backlog is large it can lag significantly behind the current event which can be a bit unexpected and counter-intuitive.

This is the note from the [README](https://github.com/uhppoted/uhppoted-mqtt#events):

> ### Events
> 
> _uhppoted-mqtt_ publishes controller events to two topics:
> 
> - `uhppoted/gateway/events`
> - `uhppoted/gateway/events/live`
> 
> The `uhppoted/gateway/events` topic is an events _feed_ that attempts to ensure that all events are published to
> the topic at least once, including events that may have occurred while the service was offline. Events are retrieved
> in small batches so if there is a significant backlog the event published in the feed may lag significantly behind the 
> current event.
> 
> The `uhppoted/gateway/events/live` topic is an events _feed_ that publishes controller events as they are received.
> However, events that occur while the service is offline won't be published to this topic.
> 
> The topics are configurable in _uhppoted.conf_:
> ```
> …
> mqtt.topic.root = uhppoted/gateway
> mqtt.topic.events = ./events
> mqtt.topic.events.live = ./events/live
> …
> ```

The other changes to the _events_ processing are mostly internal and the _uhppoted/gateway/events_ feed behaviour should (hopefully) remain the same.
